xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+		s.fc = newADSFlowControl()
+		s.firstRequest = true
+		s.mu.Unlock()
+
+		select {
+		case <-s.streamCh:
+		default:
+		}
+		s.streamCh <- stream
+
+		// Backoff state is reset upon successful receipt at least one
+		// message from the server.
+		msgReceived := s.recv(ctx, stream)
+		if msgReceived {
+			return backoff.ErrResetBackoff
+		}
+		return nil
+	}
+	backoff.RunF(ctx, runStreamWithBackoff, s.backoff)
+}
+
+// send is a long running goroutine that handles sending discovery requests for
+// two scenarios:
+// - a new subscription or unsubscription request is received
+// - a new stream is created after the previous one failed
+func (s *Stream) send(ctx context.Context) {
+	var stream transport.StreamingCall
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case stream = <-s.streamCh:
+			if err := s.sendExisting(stream); err != nil {
+				// Send failed, clear the current stream. Attempt to resend will
+				// only be made after a new stream is created.
+				stream = nil
+				continue
+			}
+		case req, ok := <-s.requestCh.Get():
+			s.logger.Infof("easwars: handling request in send goroutine")
+			if !ok {
+				return
+			}
+			s.requestCh.Load()
+
+			typ := req.(xdsresource.Type)
+			if err := s.sendNew(stream, typ); err != nil {
+				stream = nil
+				continue
+			}
+		}
+	}
+}
+
+// sendNew attempts to send a discovery request based on a new subscription or
+// unsubscription. If there is no flow control quota, the request is buffered
+// and will be sent later. This method also starts the watch expiry timer for
+// resources that were sent in the request for the first time, i.e. their watch
+// state is `watchStateStarted`.
+func (s *Stream) sendNew(stream transport.StreamingCall, typ xdsresource.Type) error {
+	s.logger.Infof("easwars: sendNew for resource type %q", typ.TypeName())
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Ig there's no stream yet, skip the request. This request will be resent
+	// when a new stream is created. If no stream is created, the watcher will
+	// timeout (same as server not sending response back).
+	if stream == nil {
+		return nil
+	}
+
+	// If local processing of the most recently received response is not yet
+	// complete, i.e. fc.pending == true, queue this write and return early.
+	// This allows us to batch writes for requests which are generated as part
+	// of local processing of a received response.
+	state := s.resourceTypeState[typ]
+	if s.fc.pending.Load() {
+		select {
+		case state.bufferedRequests <- struct{}{}:
+		default:
+		}
+		s.logger.Infof("easwars: buffered type %q", typ.TypeName())
+		return nil
+	}
+
+	names := resourceNames(state.subscribedResources)
+	s.logger.Infof("easwars: sendNew for resource type %q, names: %v", typ.TypeName(), names)
+	if err := s.sendMessageLocked(stream, names, typ.TypeURL(), state.version, state.nonce, nil); err != nil {
+		return err
+
+	}
+	select {
+	case <-state.bufferedRequests:
+	default:
+	}
```

Comment:
I don't get these drains for here and sendExisting. The processing of the buffered requests loops through all resource types? Does that mean we can drain this when we send a request for one type? I guess sendExisting sends for all so that makes more sense.

Reply:
Oh whoops I missed that bufferedRequests is scoped to resource type, I thought it was all the buffered requests for the whole stream.